pkg/reconciler/topology/partitionset_reconcile.go

```diff
@@ -0,0 +1,278 @@
+		if oldPartition.Spec.Selector != nil {
+			oldMatchExpressions = oldPartition.Spec.Selector.MatchExpressions
+		}
+		if !equality.Semantic.DeepEqual(oldMatchExpressions, newMatchExpressions) {
+			if err := c.deletePartition(ctx, logicalcluster.From(oldPartition).Path(), oldPartition.Name); err != nil && !apierrors.IsNotFound(err) {
+				conditions.MarkFalse(
+					partitionSet,
+					topologyv1alpha1.PartitionsReady,
+					topologyv1alpha1.ErrorGeneratingPartitionsReason,
+					conditionsv1alpha1.ConditionSeverityError,
+					"old partition could not get deleted",
+				)
+				return err
+			}
+			continue
+		}
+
+		// MatchLabels need to be the same
+		partitionKey := ""
+		oldMatchLabels := map[string]string{}
+		if oldPartition.Spec.Selector != nil {
+			oldMatchLabels = oldPartition.Spec.Selector.MatchLabels
+		}
+
+		// Sorting the keys of the old partition for consistent comparison
+		sortedOldKeys := make([]string, len(oldMatchLabels))
+
+		i := 0
+		for k := range oldMatchLabels {
+			sortedOldKeys[i] = k
+			i++
+		}
+		sort.Strings(sortedOldKeys)
+		for _, key := range sortedOldKeys {
+			partitionKey = partitionKey + "+" + key + "=" + oldMatchLabels[key]
+		}
+		if _, ok := matchLabelsMap[partitionKey]; ok {
+			existingMatches[partitionKey] = placeHolder
+		} else {
+			if err := c.deletePartition(ctx, logicalcluster.From(oldPartition).Path(), oldPartition.Name); err != nil {
+				conditions.MarkFalse(
+					partitionSet,
+					topologyv1alpha1.PartitionsReady,
+					topologyv1alpha1.ErrorGeneratingPartitionsReason,
+					conditionsv1alpha1.ConditionSeverityError,
+					"old partition could not get deleted",
+				)
+				return err
+			}
+		}
+	}
+	// Create partitions when no existing partition for the set has the same selector.
+	for key, matchLabels := range matchLabelsMap {
+		if _, ok := existingMatches[key]; !ok {
+			partition := generatePartition(partitionSet.Name, newMatchExpressions, matchLabels, partitionSet.Spec.Dimensions)
+			partition.OwnerReferences = []metav1.OwnerReference{
+				*metav1.NewControllerRef(partitionSet, topologyv1alpha1.SchemeGroupVersion.WithKind("PartitionSet")),
+			}
+			_, err = c.createPartition(ctx, logicalcluster.From(partitionSet).Path(), partition)
+			if err != nil && !apierrors.IsAlreadyExists(err) {
+				conditions.MarkFalse(
+					partitionSet,
+					topologyv1alpha1.PartitionsReady,
+					topologyv1alpha1.ErrorGeneratingPartitionsReason,
+					conditionsv1alpha1.ConditionSeverityError,
+					"partition could not get created",
+				)
+				return err
+			}
+		}
+	}
+	conditions.MarkTrue(partitionSet, topologyv1alpha1.PartitionsReady)
+	return nil
+}
+
+// partition populates shard label selectors according to dimensions.
+// It only keeps selectors that have at least one match.
+func partition(shards []*corev1alpha1.Shard, dimensions []string, shardSelectorLabels map[string]string) (matchLabelsMap map[string]map[string]string) {
```

Comment:
would it make sense to rename to `partitionByDimensionsAndLabels`?

Reply:
I don't see the need for now. There is currently a single way of partitioning. `Labels`could also be ambiguous in this context. `Dimensions`are a way to slice and dice by `Labels`. `Labels`are also used for filtering, defining the boundaries of the piece to slice and dice. What I mean is that `Dimensions`and `Labels`are not of the same nature for the partitioning.